crates/net/discv5/src/lib.rs

```diff
@@ -551,6 +388,170 @@ pub struct DiscoveredPeer {
+                builder.ip4(ip);
+            }
+            builder.udp4(port);
+            builder.tcp4(*tcp_port);
+
+            (IpMode::Ip4, (ip, port).into())
+        }
+        ListenConfig::Ipv6 { ip, port } => {
+            if ip != Ipv6Addr::UNSPECIFIED {
+                builder.ip6(ip);
+            }
+            builder.udp6(port);
+            builder.tcp6(*tcp_port);
+
+            (IpMode::Ip6, (ip, port).into())
+        }
+        ListenConfig::DualStack { ipv4, ipv4_port, ipv6, ipv6_port } => {
+            if ipv4 != Ipv4Addr::UNSPECIFIED {
+                builder.ip4(ipv4);
+            }
+            builder.udp4(ipv4_port);
+            builder.tcp4(*tcp_port);
+
+            if ipv6 != Ipv6Addr::UNSPECIFIED {
+                builder.ip6(ipv6);
+            }
+            builder.udp6(ipv6_port);
+
+            (IpMode::DualStack, (ipv6, ipv6_port).into())
+        }
+    };
+
+    // identifies which network node is on
+    let network_stack_id = fork.as_ref().map(|(network_stack_id, fork_value)| {
+        builder.add_value_rlp(network_stack_id, alloy_rlp::encode(fork_value).into());
+        *network_stack_id
+    });
+
+    // add other data
+    for (key, value) in other_enr_kv_pairs {
+        builder.add_value_rlp(key, value.clone().into());
+    }
+
+    // enr v4 not to get confused with discv4, independent versioning enr and
+    // discovery
+    let enr = builder.build(sk).expect("should build enr v4");
+
+    // backwards compatible enr
+    let bc_enr = NodeRecord::from_secret_key(socket, sk);
+
+    (enr, bc_enr, network_stack_id, ip_mode)
+}
+
+/// Bootstraps underlying [`discv5::Discv5`] node with configured peers.
+pub async fn bootstrap(
+    bootstrap_nodes: HashSet<BootNode>,
+    discv5: &Arc<discv5::Discv5>,
+) -> Result<(), Error> {
+    trace!(target: "net::discv5",
+        ?bootstrap_nodes,
+        "adding bootstrap nodes .."
+    );
+
+    let mut enr_requests = vec![];
+    for node in bootstrap_nodes {
+        match node {
+            BootNode::Enr(node) => {
+                if let Err(err) = discv5.add_enr(node) {
+                    return Err(Error::AddNodeFailed(err))
+                }
+            }
+            BootNode::Enode(enode) => {
+                let discv5 = discv5.clone();
+                enr_requests.push(async move {
+                    if let Err(err) = discv5.request_enr(enode.to_string()).await {
+                        debug!(target: "net::discv5",
+                            ?enode,
+                            %err,
+                            "failed adding boot node"
+                        );
+                    }
+                })
+            }
+        }
+    }
+
+    // If a session is established, the ENR is added straight away to discv5 kbuckets
+    Ok(_ = join_all(enr_requests).await)
+}
+
+/// Backgrounds regular look up queries, in order to keep kbuckets populated.
+pub fn spawn_populate_kbuckets_bg(
```

Comment:
hmm, not sure we want to expose this because this spawns something that should only be done on start?

Reply:
users are welcome to find other use cases for it imo :)